db/version_set.cc

```diff
@@ -1100,9 +1102,13 @@ class LevelIterator final : public InternalIterator {
   }
 
   void Prepare(const std::vector<ScanOptions>* scan_opts) override {
+    // We assume here that scan_opts is sorted such that
+    // scan_opts[0].range.start < scan_opts[1].range.start
     scan_opts_ = scan_opts;
-    if (file_iter_.iter()) {
-      file_iter_.Prepare(scan_opts_);
+    if (scan_opts_ != nullptr) {
+      for (size_t i = 0; i < flevel_->num_files; i++) {
+        PrepareFileIterator(i);
```

Comment:
It'll be wasteful to call PrepareFileIterator() for every file in the level. We can do an overlap check and only call it if atleast one scan_opt overlaps (and let PrepareFileIterator figure out the full set of overlapping scan_opts).